File: mod1/code/cdf.c
#include <stdio.h>
#include <stdlib.h>

// To compile the code with GCC, use the -std=gnu2x or -std=c2x flag:
// gcc -std=gnu2x -o dfp_example dfp_example.c

int main() {
    // Declare decimal floating-point variables
    _Decimal128 d128 = 1234567890123456789012345678901234.0DL;
    _Decimal64 d64 = 1234567890123456.0DD;
    _Decimal32 d32 = 1234567.0DF;

    // Print decimal floating-point values
    printf("Decimal128: %Lf\n", (long double)d128); // Note: Printing requires casting
    printf("Decimal64: %lf\n", (double)d64);
    printf("Decimal32: %f\n", (float)d32);

    return 0;
}

File: mod1/code/cdf.py
from decimal import Decimal, getcontext

"""
**Comparison: C vs. Python

| Feature                | C (IEEE 754-2008)                     | Python (`decimal` module)             |
|------------------------|---------------------------------------|---------------------------------------|
| **Types**              | `_Decimal128`, `_Decimal64`, `_Decimal32` | `Decimal` (arbitrary precision)       |
| **Precision**          | Fixed (34, 16, 7 decimal digits)      | Arbitrary (set by `getcontext().prec`)|
| **Ease of Use**        | Requires compiler support and casting | Easy to use, part of the standard library |
| **Portability**        | Limited to compilers supporting DFP   | Works on all Python implementations   |
| **Use Case**           | Low-level systems programming         | High-level applications, financial calculations |
"""

# Set the precision (number of significant digits)
getcontext().prec = 50

# Create decimal floating-point numbers
d1 = Decimal('1234567890123456789012345678901234.0')
d2 = Decimal('1234567890123456.0')
d3 = Decimal('1234567.0')

# Perform arithmetic operations
result = d1 + d2 * d3

# Print results
print("d1:", d1)
print("d2:", d2)
print("d3:", d3)
print("Result of d1 + d2 * d3:", result)